File: Editorial/0059_Spiral_Matrix_II/Go/1_Traverse_Layer_by_Layer_in_Spiral_Form.go
func generateMatrix(n int) [][]int {
    result := make([][]int, n)
    for i := range result {
        result[i] = make([]int, n)
    }
    cnt := 1
    for layer := 0; layer < (n+1)/2; layer++ {
        // direction 1 - traverse from left to right
        for ptr := layer; ptr < n-layer; ptr++ {
            result[layer][ptr] = cnt
            cnt++
        }
        // direction 2 - traverse from top to bottom
        for ptr := layer + 1; ptr < n-layer; ptr++ {
            result[ptr][n-layer-1] = cnt
            cnt++
        }
        // direction 3 - traverse from right to left
        for ptr := n - layer - 2; ptr >= layer; ptr-- {
            result[n-layer-1][ptr] = cnt
            cnt++
        }
        // direction 4 - traverse from bottom to top
        for ptr := n - layer - 2; ptr > layer; ptr-- {
            result[ptr][layer] = cnt
            cnt++
        }
    }
    return result
}

File: Editorial/59_Spiral_Matrix_II/TypeScript/1_Traverse_Layer_by_Layer_in_Spiral_Form.ts
function generateMatrix(n: number): number[][] {
    let result = Array.from({ length: n }, () => Array(n).fill(0));
    let cnt = 1;
    for (let layer = 0; layer < Math.floor((n + 1) / 2); layer++) {
        // direction 1 - traverse from left to right
        for (let ptr = layer; ptr < n - layer; ptr++) {
            result[layer][ptr] = cnt++;
        }
        // direction 2 - traverse from top to bottom
        for (let ptr = layer + 1; ptr < n - layer; ptr++) {
            result[ptr][n - layer - 1] = cnt++;
        }
        // direction 3 - traverse from right to left
        for (let ptr = n - layer - 2; ptr >= layer; ptr--) {
            result[n - layer - 1][ptr] = cnt++;
        }
        // direction 4 - traverse from bottom to top
        for (let ptr = n - layer - 2; ptr > layer; ptr--) {
            result[ptr][layer] = cnt++;
        }
    }
    return result;
}
